lib/OpenQA/Task/Git/Clone.pm

```diff
@@ -112,10 +118,15 @@ sub _git_clone ($job, $ctx, $path, $url) {
         $ctx->debug(qq{Remote default branch $remote_default});
     }
 
-    my $current_branch = _get_current_branch($path);
-    # updating default branch (including checkout)
-    _git_fetch($path, $requested_branch);
-    _git_reset_hard($path, $requested_branch) if ($requested_branch eq $current_branch);
+    if (_git_is_workdir_clean($path)) {
```

Comment:
We should try to do things as early as possible to avoid unnecessary git calls.
I would move that up before the `_get_remote_default_branch` and just return. Then we also save one nesting level.